### PR TITLE
fix(deepseek_v4): numerical stability, mask padding, and tokenizer loading

### PR DESCRIPTION
**Pull request by eauchs**
Hey @Blaizzy ! Here are the fixes for the issues raised by @machiabeli and some testing bugs we encountered:

1. **Numerical Stability:** Replaced `nn.softplus` with `mx.logaddexp(scores, mx.zeros_like(scores))` in `_score_func` for `sqrtsoftplus`. This prevents the FP16/BF16 overflow that causes NaNs in the MoE routing.
2. **Compressed Attention Mask:** The padding in `create_attention_mask` for compressed KV rows was using `mx.ones` (which translates to blocking attention). Switched it to `mx.zeros` so the model actually attends to its compressed cache instead of ignoring it.
3. **Tokenizer Fallback:** Restored the fallback to `PreTrainedTokenizerFast` in `tokenizer_utils.py` because Transformers rejects the `deepseek_v4` architecture string and throws an error on initialization.

With these changes, the model correctly accumulates in the `DeepseekV4Cache` and the throughput on Flash V4 jumps significantly during decode.

### CODE REVIEW

**Comment by eauchs** [comment]
Hey @Blaizzy — I spotted the three issues flagged by @machiabeli in his review of #1192 and patched them here:

sqrtsoftplus overflow: nn.softplus(x) overflows for large scores in BF16. Replaced with mx.logaddexp(scores, mx.zeros_like(scores)) (log-sum-exp stable).
Compressed attention mask: padding with mx.ones was blocking attention to compressed KV rows instead of allowing it. Replaced with mx.zeros.
Tokenizer fallback: added PreTrainedTokenizerFast fallback for unregistered model configs (deepseek_v4 not yet in transformers).

Happy to cherry-pick these onto your branch directly if that's easier than merging this PR.

**Comment by Blaizzy** [comment]
Thanks @eauchs!
All are great please do add, except the tokenizer part , because transformers already has a PR that fixes it.

**Comment by Blaizzy** [comment]
Please cherry pick and add here 👌🏽

**Comment by eauchs** [comment]
Done! Updated the PR and removed the tokenizer changes as requested. Ready for merge

File: mlx_lm/models/deepseek_v4.py [hunk 1]
```diff
@@ -1086,6 +1086,10 @@ def sanitize(self, weights: Dict[str, mx.array]) -> Dict[str, mx.array]:
         for k, v in weights.items():
             if k.startswith("mtp."):
                 continue
+            if ".hc_attn." in k:
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: Are you sure about this change? It looks like it will break because I kept the keys as is.

File: mlx_lm/tokenizer_utils.py
Comment (Blaizzy) [inline mlx_lm/tokenizer_utils.py:1]: Please revert this, because transformers already fixes it 

So if you install this branch it should work

https://github.com/huggingface/transformers/pull/45616

File: mlx_lm/models/deepseek_v4.py [hunk 2]
```diff
@@ -97,7 +97,7 @@ def _score_func(scores: mx.array, func: str) -> mx.array:
     if func == "sigmoid":
         return mx.sigmoid(scores)
     if func == "sqrtsoftplus":
-        return mx.sqrt(nn.softplus(scores))
+        return mx.sqrt(mx.logaddexp(scores, mx.zeros_like(scores)))
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py:100]: Please add the guy as co-author in your commit 👌🏽

**Review by Blaizzy — CHANGES_REQUESTED** [review]
Overall LGTM, just a couple nits.

File: mlx_lm/models/deepseek_v4.py (hunk 1 above)
Comment (Blaizzy) [reply mlx_lm/models/deepseek_v4.py]: Please revert this

**Comment by Blaizzy** [comment]
Two things 

Sanitize changes to revert and adding the guy you cherry picked from as co-author to the commits

**Comment by eauchs** [comment]
Updated — reverted the HC key remapping and tokenizer changes, added machiabeli as co-author. Ready for merge

**Comment by Blaizzy** [comment]
One more thing 

Please revert the renaming on hc and keep only these two changes 
image

**Comment by eauchs** [comment]
Reverted the HC renaming — back to attn_hc/ffn_hc with the sanitize remapping. Only the two validated fixes remain (logaddexp + mx.zeros). Ready for merge bro

**Comment by Blaizzy** [comment]
By revert, I mean please remove the hc renaming.

The only changes should be line 100 and 923.

**Comment by Blaizzy** [comment]
Then we can merge

**Comment by eauchs** [comment]
Done — only lines 100 and 923 changed. Ready for merge

**Review by Blaizzy — APPROVED** [review]
LGTM, thanks! 🚀
